### PR TITLE
CSHARP-5241: Generate more optimal MQL for Join using pipeline argument to $lookup.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Expressions/AstExpression.cs [hunk 1]
```diff
@@ -961,7 +961,21 @@ private static IEnumerable<AstExpression> FlattenNaryArgs(IEnumerable<AstExpress
 
         public virtual string ConvertToFieldPath()
         {
-            throw new InvalidOperationException();
+            throw new InvalidOperationException($"{this} cannot be converted to a field path.");
+        }
+
+        public string ConvertToDottedFieldNameIfPossible()
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Expressions/AstExpression.cs:967]: FYI the difference between a "field path" and a "dotted field name" is that a "field path" always starts with one or two dollar signs.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Stages/AstLookupStage.cs [hunk 2]
```diff
@@ -16,108 +16,32 @@
 using MongoDB.Bson;
 using MongoDB.Driver.Core.Misc;
 using MongoDB.Driver.Linq.Linq3Implementation.Ast.Visitors;
-using MongoDB.Driver.Linq.Linq3Implementation.Misc;
-using System.Collections.Generic;
-using System.Linq;
 
 namespace MongoDB.Driver.Linq.Linq3Implementation.Ast.Stages
 {
-    internal abstract class AstLookupStageMatch : AstNode
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Stages/AstLookupStage.cs:25]: These classes have been removed.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Stages/AstLookupStage.cs [hunk 3]
```diff
@@ -16,108 +16,32 @@
...
-            if (pipeline == _pipeline && let == _let)
-            {
-                return this;
-            }
-
-            return new AstLookupStageUncorrelatedMatch(pipeline, let);
-        }
-    }
-
     internal sealed class AstLookupStage : AstStage
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Stages/AstLookupStage.cs:22]: This class has been refactored into three classes for clarity:

1. AstLookupStage
2. AstCorrelatedSubqueryLookupStage
3. AstConciseCorrelatedSubqueryLookupStage

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/LambdaExpressionExtensions.cs [hunk 4]
```diff
@@ -24,7 +24,14 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Misc
...
     {
-        public static string GetFieldPath(this LambdaExpression fieldSelectorLambda, TranslationContext context, IBsonSerializer parameterSerializer)
+        public static bool LambdaBodyReferencesParameter(this LambdaExpression lambda, ParameterExpression parameter)
+        {
+            var visitor = new ExpressionIsReferencedVisitor(parameter);
+            visitor.Visit(lambda.Body);
+            return visitor.ExpressionIsReferenced;
+        }
+
+        public static string TranslateToDottedFieldName(this LambdaExpression fieldSelectorLambda, TranslationContext context, IBsonSerializer parameterSerializer)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/LambdaExpressionExtensions.cs:34]: Renamed to make a distinction between "field path" (which always begins with one or two dollar signs) and "dotted field name" which does not start with a dollar sign.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4054Tests.cs [hunk 5]
```diff
@@ -44,14 +45,27 @@ join movie in movies.AsQueryable() on movieId equals movie.Id
...
-                "{ $unwind : '$_inner' }",
-                "{ $project : { person : '$_outer.person', movie : '$_inner', _id : 0 } }");
+            var targetWireVersion = CoreTestConfiguration.MaxWireVersion;
+            if (Feature.ConciseCorrelatedSubqueries.IsSupported(targetWireVersion))
+            {
+                AssertStages(
+                    stages,
+                    "{ $project : { _v : { $map : { input : '$MovieIds', as : 'movieId', in : { person : '$$ROOT', movieId : '$$movieId' } } }, _id : 0 } }",
+                    "{ $unwind : '$_v' }",
+                    "{ $lookup : { from : 'movies', localField : '_v.movieId', foreignField : '_id', let : { outer : '$$ROOT' }, pipeline : [{ $project : { person : '$$outer._v.person', movie  : '$$ROOT', _id : 0 } }], as : '_v' } }",
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4054Tests.cs:55]: When the projection only uses part of `$$outer` the repetition of outer in each pipeline result is not so onerous.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationWithLinq2Tests/MongoQueryableTests.cs [hunk 6]
```diff
@@ -667,12 +667,23 @@ public void Join_method()
...
-                "{ $project : { _outer : '$$ROOT', _id : 0 } }",
-                "{ $lookup : { from : 'testcollection_other', localField : '_outer._id', foreignField : '_id', as : '_inner' } }",
-                "{ $unwind : '$_inner' }",
-                "{ $project : { p : '$_outer', o : '$_inner', _id : 0 } }");
+            var targetWireVersion = CoreTestConfiguration.MaxWireVersion;
+            if (Feature.ConciseCorrelatedSubqueries.IsSupported(targetWireVersion))
+            {
+                Assert(query,
+                    1,
+                    "{ $lookup : { from : 'testcollection_other', localField : '_id', foreignField : '_id', let : { outer : '$$ROOT' }, pipeline : [{ $project : { p : '$$outer', o : '$$ROOT', _id : 0 } }], as : '_v' } }",
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationWithLinq2Tests/MongoQueryableTests.cs:675]: When the entire `$$outer` is included in each pipeline result there is a lot of data duplication...

The main point of this PR was to support larger inner result sets, but as this query shows that isn't always true. It's only true when an individual result is smaller than an individual inner document.

Let's discuss...

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationWithLinq2Tests/MongoQueryableTests.cs [hunk 7]
```diff
@@ -702,12 +724,23 @@ public void Join_syntax()
...
-                "{ $project : { _outer : '$$ROOT', _id : 0 } }",
-                "{ $lookup : { from : 'testcollection_other', localField : '_outer._id', foreignField : '_id', as : '_inner' } }",
-                "{ $unwind : '$_inner' }",
-                "{ $project : { A : '$_outer.A', CEF : '$_inner.CEF', _id : 0 } }");
+            var targetWireVersion = CoreTestConfiguration.MaxWireVersion;
+            if (Feature.ConciseCorrelatedSubqueries.IsSupported(targetWireVersion))
+            {
+                Assert(query,
+                    1,
+                    "{ $lookup : { from : 'testcollection_other', localField : '_id', foreignField : '_id', let : { outer : '$$ROOT' }, pipeline : [{ $project : { A : '$$outer.A', CEF : '$CEF', _id : 0 } }], as : '_v' } }",
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationWithLinq2Tests/MongoQueryableTests.cs:732]: This is an example of a query that benefits from this PR. The individual results of `pipeline` are smaller than an entire inner document.

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**